### PR TITLE
Fix cross-compiling i686-pc-windows-gnu from Linux

### PR DESCRIPTION
**Pull request by Manishearth**
This is a patch from @neersighted from the Tor project, which lets Tor
cross compile on Windows.

See the commit message for details.

https://trac.torproject.org/projects/tor/ticket/28157

I don't actually understand this patch (see the commit message for an
explanation), I'm just helping upstream it.

r? @alexcrichton @Mark-Simulacrum

### CODE REVIEW

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -150,6 +150,11 @@ pub fn std_cargo(builder: &Builder,
         cargo.env("MACOSX_DEPLOYMENT_TARGET", target);
     }
 
+    // FIXME: Temporary detection of SJLJ MinGW compilers.
+    if build.build.build.contains("linux") && target == "i686-pc-windows-gnu" {
```
Comment (phansch) [inline src/bootstrap/compile.rs]: Should this be `builder.build.build`?

**Comment by neersighted** [comment]
@alexcrichton can correct me if I'm wrong, but I originally discontinued this patch since I realized it wouldn't work as-is for officially distributed toolchains; targets seem to be shared between hosts, so this patch will not quite work as it depends on compile-time checks.

Instead a different approach is likely needed; this PR could likely be converted to use runtime detection.

I might take another look making this change if review finds this change solves the issue (when compiled on 32-bit Linux), as I first researched and tested this issue many months ago and my memory of it is a bit fuzzy.

**Comment by alexcrichton** [comment]
Ah yes indeed @neersighted! I think in addition to this causing discrepancy across hosts I think that this largely just builds rather than works right? I think that with this patch it means that panics won't actually run Rust destructors because we've configured LLVM to emit dwarf information rather than use sj/lj for exception handling.

@neersighted does that sound right?

**Comment by alexcrichton** [comment]
Ok! In that case I think we probably shouldn't land this PR, for the primary reason that while this produces linked binaries it doesn't actually produce correct binaries. The incorrectness specifically stems from the fact that LLVM is, for i686-pc-windows-gnu, always generating DWARF unwinding tables. This means that if the sjlj runtime for throwing exceptions is used instead that when a panic happens no destructors will be run.

I think there's probably two fixes here, both of which are unfortunately quite difficult:

* For tor's use case specifically, somehow (I'm not sure how myself) acquire a mingw compiler that uses dwarf for exception handling. This will be compatible with the `i686-pc-windows-gnu` target as-is today.
* Otherwise add an entirely new target, like `i686-pc-windows-gnu-sjlj` (or something like that). This entire target would be geared towards sjlj exception handling, and this would involve changing how we generate code in LLVM as well as including the runtime support changes in this patch itself.

I think we probably have basically no path to getting `i686-pc-windows-gnu` as-is today working with the mingw toolchains on linux which use sjlj exception handling. Even if we could change LLVM at runtime the binary of the standard library would be in one mode or the other.

**Comment by neersighted** [comment]
I do think there is a third, (slightly) easier path, though I'm not sure that it is in line with Rust's philosophy and compiler design. The compiler could be made to use multiple unwinding formats, and pick from them at build-time. Essentially, pushing the decision on what to link to from the standard library to the compiler (which also would have to emit different unwinding tables).

However, this couples the compiler much more closely to the target environment itself, and could potentially be problematic when building for targets that do not use the standard library (depending on how it was implemented), like ARM uCs.

**Comment by malbarbo** [comment]
[cross](https://github.com/rust-embedded/cross) [builds](https://github.com/rust-embedded/cross/blob/master/docker/mingw.sh) a `i686-pc-windows-gnu` that uses dwarf for exception handling.

**Comment by alexcrichton** [comment]
@neersighted oh I think that's basically the same route as taking a new target. We can add a compiler flag but it's not too useful unless the standard library is recompiled, which is effectively a new target anyway.

@malbarbo oh that's perfect! Thanks for the link! I'll send that to the tor folks as well

**Comment by neersighted** [comment]
Well, the key difference is (weak?) linking in the standard library and the compiler selecting the appropriate symbols instead of the standard library's feature detection. Still, it would require intrusive changes to avoid another target, so I concede it is not the best option.

I too was unaware of mingw builds from cross... Are they capable of unwinding across SEH frames (DWARF2 is not capable of crossing SEH as far as I know)? Are SEH frames (built by MSVC most likely) ever present in rust's call stack when cross-compiling to Windows?

**Comment by alexcrichton** [comment]
@neersighted oh but it's not just about the symbols used, right? It's all about how the compiler actually compiles libraries? We ship a binary version of the standard library to users, and while we could dynamically pick the right symbol at runtime the binary itself is only compatible with one strategy (not the other)

(I'm not sure about SEH and mingw, I don't know how the two interact)

**Comment by superriva** [comment]
So, globally:

1. Rust's toolchains support dwarf exceptions for win32 and seh exceptions for win64.
2. Rust's toolchains doesn't have mingw compilers internally, so developers have to install own mingw compilers into their systems. And these local mingw compilers can support different exception mode. For debian\ubuntu are: sjls for win32 (it's information from the debian\ubuntu maintainer of mingw-w64) and seh for win64.
3. For right working both mingw compilers and Rust's toolcahins must support similar exception mode.
4. malbarbo's patch shows how to rebuild local mingw compiler for dwarf exceptions.
5. torproject's patch shows how to build Rust's toolchains for sljl exceptions.

Right? Thx.

So, split i686-pc-windows-gnu into two new toolchains like i686-pc-windows-gnu-dwarf and i686-pc-windows-gnu-sjlj is a pragmatical idea.

**Comment by neersighted** [comment]
@alexcrichton yeah, my idea was to ship a 'fat' version of the standard library essentially, that includes both. But upon digging into the code, it looks like it is significantly easier to add a new target, which should likely be the way forward.

@superriva that's a somewhat accurate summation, but there are additional considerations with compatibility between unwinding formats. As far as I know, 32-bit MinGW GCC uses SJLJ because DWARF2 cannot traverse SEH frames, which are what MSVC outputs. 64-bit MinGW GCC just emits SEH.

My patch merely links against (internal) SJLJ symbols, but does not configure LLVM itself to use SJLJ.

File: src/bootstrap/compile.rs (hunk 1 above)
Comment (superriva) [reply src/bootstrap/compile.rs]: target.contains("linux") ?

as like:
...
766 !target.contains("freebsd") &&
767 !target.contains("windows") &&
768 !target.contains("apple") {
...

**Pull request closed without merging**